feat(eslint-plugin): [prefer-return-this-type] add a new rule

Zzzen opened the pull request:
close #889
The implementation is pretty clear, it only warns if:
1. the return type is written explicitly, and it is simple ( not union or intersection types )
2. all code paths return `this` type

TODO:
- [x] doc
- [x] maybe support PropertyDeclaration initialized with functions like
```ts
class Foo {
  f = (): Foo => {
    return this;
  }
}
```

bradzacher commented on packages/eslint-plugin/docs/rules/prefer-return-this-type.md:
```diff
@@ -0,0 +1,83 @@
+  f2() {
+    return this;
+  }
+  f3(): Foo | undefined {
```
why not `this | undefined`?

Zzzen replied:
👌

bradzacher commented on packages/eslint-plugin/src/rules/prefer-return-this-type.ts:
```diff
@@ -0,0 +1,161 @@
+      description:
+        'Enforce that `this` is used when only `this` type is returned',
+      category: 'Best Practices',
+      recommended: 'error',
```
adding to the recommended set is a breaking change

```suggestion
      recommended: false,
```

bradzacher replied:
adding to the recommended set is a breaking change

```suggestion
      recommended: false,
```

bradzacher commented on packages/eslint-plugin/src/rules/prefer-return-this-type.ts:
```diff
@@ -0,0 +1,161 @@
+      requiresTypeChecking: true,
+    },
+    messages: {
+      UseThisType: 'use `this` type instead.',
```
nit - convention is to use camelCase
```suggestion
      useThisType: 'use `this` type instead.',
```

bradzacher commented on packages/eslint-plugin/src/rules/prefer-return-this-type.ts:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    return {
+      'ClassBody > MethodDefinition'(node: TSESTree.MethodDefinition): void {
+        checkFunction(node.value, node.parent!.parent as ClassLikeDeclaration);
+      },
+      'ClassBody > ClassProperty'(node: TSESTree.ClassProperty): void {
+        if (
+          !(
+            node.value?.type === AST_NODE_TYPES.FunctionExpression ||
+            node.value?.type === AST_NODE_TYPES.ArrowFunctionExpression
+          )
+        ) {
+          return;
+        }
+
+        checkFunction(node.value, node.parent!.parent as ClassLikeDeclaration);
+      },
```
one easier way to do this is to use a stack to track your state.
It saves you doing manual traversal of the tree

```ts
interface Stack { classLike: ClassLikeDeclaration, methodLike: FunctionLike | null }
const stack: Stack[] = [];
function getCurrentStack(): Stack | undefined {
  return stack[stack.length - 1];
}

return {
  'ClassDeclaration, ClassExpression:enter'(node: ClassLikeDeclaration) {
    stack.push({ classLike: node, methodLike: null });
  },
  'ClassDeclaration, ClassExpression:exit'(node: ClassLikeDeclaration) {
    stack.pop();
  },
  'ClassBody > MethodDefinition:enter, ClassBody > ClassProperty > :matches(FunctionExpression, ArrowFunctionExpression).value:enter'(node: FunctionLike ) {
    const current = getCurrentStack();
    if (current != null) {
      current.methodLike = node;
    }
  },
  'ClassBody > MethodDefinition, ClassBody > ClassProperty > :matches(FunctionExpression, ArrowFunctionExpression).value:exit'(node: FunctionLike ) {
    const current = getCurrentStack();
    if (current != null) {
      current.methodLike = null;
    }
  },
  ReturnStatement(node) {
    const current = getCurrentStack();
    if (current != null && current.methodLike != null) {
      // check return and report
    }
  },
};
```

Zzzen replied:
Cool, I haven't seen this pattern before, will try it.

Zzzen replied:
This is more complicated than I have expected. We may have functions inside methods and need to track them too.

```ts
class Foo {
  bar() {
    const f = function() {
       return this;
    }
  }
}
```

bradzacher reviewed (changes requested):
a few quick comments - haven't reviewed SUPER deeply yet.

---

what about `abstract` methods? Do we want to check those signatures?
what about `interface`s? Do we want to check those signatures?

Zzzen commented:
>what about abstract `methods`? Do we want to check those signatures?
what about `interfaces`? Do we want to check those signatures?


Showerthought: we don't need to check them because `this` doesn't make a difference in `interfaces`/`methods`. Eg. [Playgroud](https://www.staging-typescript.org/play?#code/JYOwLgpgTgZghgYwgAgJIEETALZwDYBiwAHhACbIDeAUMnctgPYBuEAFAJQBcamO+RUmQDc1AL7VqoSLEQoMWXHgAKjPAE8mABwAWwBFVr0mrTjzB6AzqInUEeOJcvIAwnDCDyyHFrwRsEODOCvyEJF409MgA9NHIUBBgAK5QIMhg6looAO7AeHjIAEYoCDoQCADWXoXq6VbxSX5GdCbs3HXAzpFR9AnJqR3Wzci2UQGM2WaDhj29iSlpFp2iURK29o7ObmCqGtp6Bj5+AUG8ivi7moy6+jPGLG3m9d2zfQuDK-SjxhATU0tdYZRN4DAGfOhrSQwJIgBBgYCMNJJSwQVDgaDwJBsGDhMg8EJKTxkAA0yC0aiuNwQ+L4Sku+30HDudFi8XmoMyEBp5zCQmGOKEADpWpxRMNWSDFpzuaEicNyXtrgdhQ9ReIoTC4QikSjUNhfP5AmB3NrsbieNsiaSFZSDhb3PSlYzmTE4pL0tLXO45VEBeQVaYOGKohL2VKsvadhSGQh5dGnQgA20bNQgA)
```ts
interface IAnimalFixed {
    move(): IAnimalFixed;
}

interface IAnimalPolymophic {
    move(): this;
}

class CatFixed implements IAnimalFixed {
    // return type will be checked by this rule
    move(): this {
        return this;
    }

    meow(): this {
        return this;
    }
}

class CatPolymophic implements IAnimalPolymophic {
    move(): this {
        return this;
    }

    meow(): this {
        return this;
    }
}

function useInterface(fixed: IAnimalFixed, polymophic: IAnimalPolymophic) {
    // return type: IAnimalFixed
    fixed.move();

    // return type: IAnimalFixed
    polymophic.move();
}

function useImplementation(fixed: CatFixed, polymophic: CatPolymophic) {
    // return type: CatFixed
    fixed.move();

    // return type: CatPolymophic
    polymophic.move();
}



```

bradzacher commented on packages/eslint-plugin/src/configs/recommended-requiring-type-checking.ts:
```diff
@@ -17,6 +17,7 @@ export = {
     '@typescript-eslint/no-unsafe-member-access': 'error',
     '@typescript-eslint/no-unsafe-return': 'error',
     '@typescript-eslint/prefer-regexp-exec': 'error',
+    '@typescript-eslint/prefer-return-this-type': 'error',
```
you need to regenerate the configs

bradzacher commented on packages/eslint-plugin/README.md:
```diff
@@ -163,6 +163,7 @@ Pro Tip: For larger codebases you may want to consider splitting our linting int
 | [`@typescript-eslint/prefer-readonly-parameter-types`](./docs/rules/prefer-readonly-parameter-types.md)               | Requires that function parameters are typed as readonly to prevent accidental mutation of inputs                        |                    |          | :thought_balloon: |
 | [`@typescript-eslint/prefer-reduce-type-parameter`](./docs/rules/prefer-reduce-type-parameter.md)                     | Prefer using type parameter when calling `Array#reduce` instead of casting                                              |                    | :wrench: | :thought_balloon: |
 | [`@typescript-eslint/prefer-regexp-exec`](./docs/rules/prefer-regexp-exec.md)                                         | Enforce that `RegExp#exec` is used instead of `String#match` if no global flag is provided                              | :white_check_mark: | :wrench: | :thought_balloon: |
+| [`@typescript-eslint/prefer-return-this-type`](./docs/rules/prefer-return-this-type.md)                               | Enforce that `this` is used when only `this` type is returned                                                           | :heavy_check_mark: | :wrench: | :thought_balloon: |
```
you need to regen the docs to  remove the recommended tick

Zzzen commented:
Why is CI not running?

bradzacher reviewed (approved):
LGTM - this is enough coverage for a v1.
We can always add more cases later.
Thanks for your contribution!
